### PR TITLE
Make resize an AOT app, and reschedule it

### PR DESCRIPTION
**Pull request by abadams**
Having a good resample is something that has come up repeatedly. I rewrote resample to be an AOT app, and did some scheduling work on it. Benchmarks on my machine:

./bin/resize ../images/rgb_small.png ./bin/out_box_up.png -f 4.0 -t box
Loading '../images/rgb_small.png'
Resampling '../images/rgb_small.png' from 192x320 to 768x1280
Time: 0.322900 ms
./bin/resize ../images/rgb_small.png ./bin/out_linear_up.png -f 4.0 -t linear
Loading '../images/rgb_small.png'
Resampling '../images/rgb_small.png' from 192x320 to 768x1280
Time: 0.388100 ms
./bin/resize ../images/rgb_small.png ./bin/out_cubic_up.png -f 4.0 -t cubic
Loading '../images/rgb_small.png'
Resampling '../images/rgb_small.png' from 192x320 to 768x1280
Time: 0.658300 ms
./bin/resize ../images/rgb_small.png ./bin/out_lanczos_up.png -f 4.0 -t lanczos
Loading '../images/rgb_small.png'
Resampling '../images/rgb_small.png' from 192x320 to 768x1280
Time: 0.928800 ms
./bin/resize ../images/rgb.png ./bin/out_box_down.png -f 0.5 -t box
Loading '../images/rgb.png'
Resampling '../images/rgb.png' from 1536x2560 to 768x1280
Time: 1.585800 ms
./bin/resize ../images/rgb.png ./bin/out_linear_down.png -f 0.5 -t linear
Loading '../images/rgb.png'
Resampling '../images/rgb.png' from 1536x2560 to 768x1280
Time: 1.683400 ms
./bin/resize ../images/rgb.png ./bin/out_cubic_down.png -f 0.5 -t cubic
Loading '../images/rgb.png'
Resampling '../images/rgb.png' from 1536x2560 to 768x1280
Time: 2.049600 ms
./bin/resize ../images/rgb.png ./bin/out_lanczos_down.png -f 0.5 -t lanczos
Loading '../images/rgb.png'
Resampling '../images/rgb.png' from 1536x2560 to 768x1280
Time: 2.748500 ms

### CODE REVIEW

File: apps/resize/Makefile [hunk 1]
```diff
@@ -1,18 +1,34 @@
...
+$(BIN)/resize_%_down.a: $(BIN)/resize_generator
+	@mkdir -p $(@D)
+	$^ -g resize -o $(BIN) -f resize_$*_down target=$(HL_TARGET)-no_runtime interpolation_type=$* upsample=false
+
+$(BIN)/runtime.a: $(BIN)/resize_generator
+	@mkdir -p $(@D)
+	$^ -r runtime -o $(BIN) target=$(HL_TARGET)
 
-.PHONY: clean
+$(BIN)/resize: resize.cpp $(BIN)/resize_box_up.a $(BIN)/resize_linear_up.a $(BIN)/resize_cubic_up.a $(BIN)/resize_lanczos_up.a $(BIN)/resize_box_down.a $(BIN)/resize_linear_down.a $(BIN)/resize_cubic_down.a $(BIN)/resize_lanczos_down.a $(BIN)/runtime.a
```
Comment (steven-johnson) [inline apps/resize/Makefile]: nit: could we break this line into shorter bits with \ please?

**Comment by steven-johnson** [comment]
We should really add a CMake builder for this too, but we can do that after this lands.

**Comment by steven-johnson** [comment]
LGTM.

Would be nice for someone to adapt this further to work on uint8 and uint16 images (as well as float32), but I'm not gonna get greedy :-)

**Comment by awakecoding** [comment]
@steven-johnson the current code only does float, right? I'd like uint8 and uint16 image support as well, but I'm glad there is now a working AOT resize

**Comment by steven-johnson** [comment]
> resize_generator.cpp:31:22: error: 'M_PI' was not declared in this scope
>     return sin(float(M_PI) * x) / x;

M_PI is nonstandard-but-common :-(

**Comment by abadams** [comment]
I added minimal support for uint8 and uint16 (just casts on the input and output). Adapting this approach to do the computation in fixed-point is interesting - e.g. normalizing the filters requires enforcing that they sum to 256, which is not achievable by normalizing them as floats and then quantizing. You need to do an error diffusion thing.

**Comment by abadams** [comment]
new output:
```
planar       box   float32  4.00  time: 0.363400 ms
packed       box   float32  4.00  time: 0.561600 ms
planar       box   float32  0.50  time: 1.611400 ms
packed       box   float32  0.50  time: 0.909000 ms
planar       box    uint16  4.00  time: 0.352000 ms
packed       box    uint16  4.00  time: 0.319400 ms
planar       box    uint16  0.50  time: 1.231000 ms
packed       box    uint16  0.50  time: 0.924900 ms
planar       box     uint8  4.00  time: 0.378400 ms
packed       box     uint8  4.00  time: 0.327100 ms
planar       box     uint8  0.50  time: 1.234900 ms
packed       box     uint8  0.50  time: 0.905500 ms
planar    linear   float32  4.00  time: 0.420900 ms
packed    linear   float32  4.00  time: 1.414500 ms
planar    linear   float32  0.50  time: 1.688300 ms
packed    linear   float32  0.50  time: 1.236300 ms
planar    linear    uint16  4.00  time: 0.418700 ms
packed    linear    uint16  4.00  time: 0.373800 ms
planar    linear    uint16  0.50  time: 1.381600 ms
packed    linear    uint16  0.50  time: 1.299900 ms
planar    linear     uint8  4.00  time: 0.405300 ms
packed    linear     uint8  4.00  time: 0.445900 ms
planar    linear     uint8  0.50  time: 1.419900 ms
packed    linear     uint8  0.50  time: 1.264600 ms
planar     cubic   float32  4.00  time: 0.658700 ms
packed     cubic   float32  4.00  time: 2.352500 ms
planar     cubic   float32  0.50  time: 2.026500 ms
packed     cubic   float32  0.50  time: 2.033900 ms
planar     cubic    uint16  4.00  time: 0.702100 ms
packed     cubic    uint16  4.00  time: 0.669900 ms
planar     cubic    uint16  0.50  time: 1.646800 ms
packed     cubic    uint16  0.50  time: 2.119700 ms
planar     cubic     uint8  4.00  time: 0.612500 ms
packed     cubic     uint8  4.00  time: 0.655400 ms
planar     cubic     uint8  0.50  time: 1.637000 ms
packed     cubic     uint8  0.50  time: 2.053200 ms
planar   lanczos   float32  4.00  time: 0.888000 ms
packed   lanczos   float32  4.00  time: 2.814000 ms
planar   lanczos   float32  0.50  time: 2.616600 ms
packed   lanczos   float32  0.50  time: 3.142100 ms
planar   lanczos    uint16  4.00  time: 0.750700 ms
packed   lanczos    uint16  4.00  time: 1.049300 ms
planar   lanczos    uint16  0.50  time: 2.337200 ms
packed   lanczos    uint16  0.50  time: 3.165300 ms
planar   lanczos     uint8  4.00  time: 0.944200 ms
packed   lanczos     uint8  4.00  time: 1.033000 ms
planar   lanczos     uint8  0.50  time: 2.296300 ms
packed   lanczos     uint8  0.50  time: 3.142700 m
```

**Comment by abadams** [comment]
test failure is transient llvm build breakage. @steven-johnson PTAAL - my last commit added a bunch of stuff

**Comment by steven-johnson** [comment]
Changes LGTM

**Pull request merged**